### ISSUE TITLE
(3.0.77) Go to containing dictionary ... doesn't

### ISSUE THREAD

**Issue by dalehenrich** [ericwinger/Jade#530]
BEFORE:
![image](https://user-images.githubusercontent.com/623951/65555940-d213d100-dee2-11e9-8745-7371aa2df378.png)
AFTER:
![image](https://user-images.githubusercontent.com/623951/65555965-e061ed00-dee2-11e9-9222-57ee86b6aa61.png)

**Comment by ericwinger** [comment]
The reason ``Go to containing dictionary`` didn't go anywhere was because the containing dictionary appears to be the empty string. On one hand, it's nice that there was no walkback. On the other, it's frustrating because it's not clear how the class's dictionary could be empty. 

It's possible the class needed an update from the server, but without a reproduction case it's going to be difficult to track down. What was done before the browser got into this state? 

Turning on logging could provide a nice 'paper trail' to follow. A reproduction case would be better. 
@LisaAlmarode  If you can get logs of this happening or a reproduction case, I'd appreciate it. I've had no luck so far.

**Comment by dalehenrich** [comment]
@ericwinger where is the code that defines the "containing dictionary"? is it in the Services code? If so, add a pointer to the place where you determine the containg dictionary and I shoul be able to provide you with a test case ... the problem reported above fails consistently, and it is in a private stone of mine:
![image](https://user-images.githubusercontent.com/623951/65917491-4db2c980-e38c-11e9-84bb-643ac08c6adc.png)
If you want to login, you have to use rogue-vpn to find my machine when I am at home working ... let me know if you can't reproduce the problem and I will do the debugging on my end

**Comment by ericwinger** [comment]
@dalehenrich  Take a look at ``RowanClassService>>setDictionary:``. See if that sheds any light.

**Comment by dalehenrich** [comment]
Note that Rowan `knows` which dictionary the class is is (**Globals**), when it creates the class definition template:
![image](https://user-images.githubusercontent.com/623951/65917815-1d1f5f80-e38d-11e9-90ec-7cceb9bcba1b.png)

**Comment by dalehenrich** [comment]
Looking at senders of `RowanClassService>>setDictionary:`, I see that both `RowanClassService>>basicRefreshFrom: ` and `RowanClassService>>minimalRefreshFrom:` **can** return the meta class and with a little bit of testing in my image, I see that `(Rowan image symbolList dictionariesAndSymbolsOf: GsQueryPredicate) first first name` produces`#'Globals'` ... which is the correct answer and `(Rowan image symbolList dictionariesAndSymbolsOf: GsQueryPredicate class)` produces `anArray( )` ... which is the wrong answer. 

Anything other than a class returns an empty list, so I would suspect that Jadeite is either passing in a non-Class `theClass` or perhaps `metaClass` is defaulting to `true`. I'm pretty sure that this happens when I have just opened up fresh project browser on the class.

Here's an example where explicitly selecting the class button creates and empty list:
![image](https://user-images.githubusercontent.com/623951/65919791-38409e00-e392-11e9-8de0-ccb607aac36e.png)

**Comment by ericwinger** [comment]
Ahh, you may have found it. Try this - If you click on the ``Class`` tab, I'll bet you don't see the dictionary. That would be a nice reproduction case.

**Comment by dalehenrich** [comment]
Take a look at the final image in my previous comment ... I did select the 'Class' button ... howver, in the original report, the 'Instance' button is clearly selected, which is why I mentioned the the `metaClass` iv might default to `true` or you might be passing in a non-class object ...

**Comment by ericwinger** [comment]
This image? The class side is selected.  ![image](https://user-images.githubusercontent.com/623951/65555940-d213d100-dee2-11e9-8745-7371aa2df378.png)

**Comment by dalehenrich** [comment]
Exactly ... and as predicted ... the containing dictionary is empty ... it shouldn't be ... at a minimum this is wrong ... have you tried connecting to my stone? then you can do whatever experiments you need to do here ... AFAICT, both class tabs are already selected ... is there another class tab I am missing?

**Comment by ericwinger** [comment]
The problem is the meta class that you pointed out. I've got a fix ready for 3.0.78 for you to try. ... Just needs tests

**Comment by dalehenrich** [comment]
... your fix won't necessarily fix the original problem (or at best it will mask the problem) as in the original bug report the **instance** tab is clearly selected ... so I think it might be an initial condition problem as well ...

**Comment by dalehenrich** [comment]
Now that I've been workin gin this project ... I've added two more packages and the problem does not reproduce (instance tab ending up with empty string for dictionary name) ... so there is clearly something else going wrong here ...

**Comment by dalehenrich** [comment]
I am suspicious of this line of code:
```smalltalk
	classOrMeta := meta == true ifTrue:[theClass class] ifFalse:[theClass].
```
if meta is nil, you will return the meta class for `theClass`, which is what I am seeing ... and the fact that you aren't using meta directly implies that in some circumstances meta will be set to a non-boolean value (`nil` perhaps) ... you've got an initialize method that sets meta to `false`, but the behavior that I'm seeing in my initial case, implies that someone is using the setter (`meta:`) to set meta to a non-boolean value --- or to `true` ... I've lost my "reproducable test case" ... so I cannot test your code for you ... so I think that unless additional work is done this bug will still be lurking around waiting to hit us again ...

Looking at all references in the ServiceClass and verifying that the setters all look good and then looking at all senders of `meta:` is probably a good idea, since your fix will just cause another part of the code to go haywire, if meta is not being handled correctly --- which appears to be the case ...

**Comment by ericwinger** [comment]
Yes, ``meta`` can legitimately be nil in some cases especially when the class service originates in Dolphin. e.g. Doing a find class doesn't yet know it's meta so it assumes false. 

fyi - None of the images you've posted on this issue show the instance tab selected with the problem issue. If you can send me a picture of the instance tab selected & the problem dictionary I'll look into it further.

**Comment by dalehenrich** [comment]
Ahh, I did not notice that ... I don't believe that I did anything other than open the project browser and select the class in question (without touching any of the class tabs) I assumed that I was looking at the instance side in my original bug report. So presumably there is another condition (closely related to my original bug report, which does not reproduce now:) where the class side of a class gets selected when I open a fresh project browser and simply select the class ...

**Comment by dalehenrich** [comment]
I see that the class tab **is sticky**:
![image](https://user-images.githubusercontent.com/623951/65924305-f5d28d80-e3a0-11e9-9ead-888eb61f9cfa.png)
... this is unexpected to me ... is this "stickiness" expected behavior?

Earlier today I wasted a fair amount of time writing code on the class side of a class when I thought I was on the instance side ... I couldn't understand why the project browser "put code on the class-side of a class" ... 

I expect to be on the instance side when I switch classes, so I guess I'm going to have to start training myself to worry about the instance/class tab whenever I am working with Jadeite, unless this is considered to be a bug by you and/or @LisaAlmarode :) ... a couple more times of wasting time will probably be enough ... 

It's a bit of a surprise, that I just found this out after working in Jadeite for a year and a half, but clearly I haven't been bit by this assumption until today:)

So... it is very possible that I was working on another class when I switched to the bad boy and I just assumed that the instance side was selected... sorry about that ...

**Comment by ericwinger** [comment]
> ... this is unexpected to me ... is this "stickiness" expected behavior?

The class tab "stickiness" is actually intended in some scenarios.

For example, if you have an open browser on a class with class tab selected then use Jadeite menu option ``New Projects Browser (Ctrl+N)``. It will open up the new browser on that class with the class tab selected. The assumption is that you want to open a new browser in the same place you are currently rather than starting from a fresh browser. 

Alternatively, if you do a ``Find Class ...`` from the console's ``Jadeite>Browse`` menu, it will open on the instance side. The assumption is that you are starting your work from scratch and most work is done on the instance side of classes. 

Yet, it's possible that there might be a bug or enhancements in this area. Since I've got a fix for the specific problem coming in 3.0.78, can we close this issue for review by @LisaAlmarode and open a new issue to examine class tab "stickiness" behavior?

**Comment by dalehenrich** [comment]
Well the case where I have observed the "stickiness" is when I switch to a different class in the same browserr ... I also did a `Find Class ...` earlier today and ended up with the class tab selected ... of course it didn't reproduce --- neither of these use cases fall under your "intentional Use case" ... so there does appear to be some more bugs in this area ...

**Comment by LisaAlmarode** [comment]
This issue got a bit muddled. The bulk of this is discussion about the selection of class or instance doesn't seem to have any problems (other than new issue #555 reported).  

The case where RowanClassService on the class side showed an empty dictionary is fixed in 3.0.79.

There were cases long ago in which some menu operations acted wonky (e.g. drag and drop moved methods to the class side), due to internal defaults to meta. It seems possible that the initial report on this case may be due to a leftover case of this kind.  However, we'd need some details as to the steps preceding to hope to reproduce this... I tried a few cases but there are too many possibilities.

Given the one issue fixed, and the large amount of digressions, if we have another case of a class's go-to-dictionary turning up empty, a new issue should be opened.
